README.md
# lock_and_cache

Lock and cache using Redis!

Most caching libraries don't do locking, meaning that >1 process can be calculating a cached value at the same time. Since you presumably cache things because they cost CPU, database reads, or money, doesn't it make sense to lock while caching?

## Quickstart

```ts
import { LockAndCache } from "lock_and_cache";

const cache = new LockAndCache();

async function computeStockQuote(symbol: string) {
  // Fetch stock price from remote source, cache for one minute.
  //
  // Calling this multiple times in parallel will only run it once the cache key
  // is based on the function name and arguments.
  return 100;
}
const stockQuote = cache.wrap({ ttl: 60 }, computeStockQuote);

// If you forget this, your process will never exit.
cache.close();
```

## Install

```sh
npm install --save lock_and_cache
```

## Theory

lock_and_cache...

1. returns cached value (if exists)
2. acquires a lock
3. returns cached value (just in case it was calculated while we were waiting for a lock)
4. calculates and caches the value
5. releases the lock
6. returns the value

As you can see, most caching libraries only take care of (1) and (4) (well, and (5) of course).

## Setup

By default, you probably want to put locks and caches in separate databases:

```sh
export CACHE_URL=redis://redis:6379/2
export LOCK_URL=redis://redis:6379/3
```

## API

We have full JSDoc for this library, but these are the highlights at the time of writing.

```ts
/**
 * Valid cache keys.
 *
 * These may only contain values that we can serialize consistently.
 */
export type CacheKey = string | number | boolean | null | CacheKey[];

/** Options that can be passed to `get` and `wrap`. */
export type GetOptions = {
  /**
   * "Time to live." The time to cache a value, in seconds.
   *
   *  Defaults to `DEFAULT_TTL`.
   */
  ttl?: number;
};

/**
 * A user-supplied function that computes a value to cache.
 *
 * This function may have an optional `displayName` property, which may be used
 * as part of the cache key if this function is passed to `wrap`.
 *
 * By default, `WorkFn` functions do not accept arguments, but if you're using
 * the `wrap` feature, you may want to supply an optional argument list.
 */
type WorkFn<T, Args extends CacheKey[] = []> = ((
  ...args: Args
) => Promise<T>) & {
  displayName?: string;
};

class LockAndCache {
  /**
   * Create a new `LockAndCache`.
   *
   * @param caches Caches to use. Defaults to `tieredCache()`.
   * @param lockClient Redis client to use for locking. Defaults to using
   * `DEFAULT_REDIS_OPTIONS`.
   */
  constructor({
    lockClient = defaultRedisLockClient(),
    cacheClient = defaultRedisCacheClient(),
  } = {});

  /**
   * Shut down this cache manager. No other functions may be called after this.
   */
  close(): void;

  /**
   * Either fetch a value from our cache, or compute it, cache it and return it.
   *
   * @param key The cache key to use.
   * @param options Cache options. `ttl` is in seconds.
   * @param work A function which performs an expensive caculation that we want
   * to cache.
   */
  async get<T>(key: CacheKey, options: GetOptions, work: WorkFn<T>): Promise<T>;

  /**
   * Given a work function, wrap it in a `cache.get`, using the function's
   * arguments as part of our cache key.
   *
   * @param options Cache options. `name` is the base name of our cache key.
   * `ttl` is in seconds, and it defaults to `DEFAULT_TTL`.
   * @param work The work function to wrap. If `options.name` is not specified,
   * either `work.displayName` or `work.name` must be a non-empty string.
   */
  wrap<T, Args extends CacheKey[] = []>(
    options: GetOptions & { name?: string },
    work: WorkFn<T, Args>
  ): WorkFn<T, Args>;
}
```

## Contributing

Please send us your pull requests!


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [6.0.0-beta.5] - 2020-08-26

### Added

- Allow objects with potentially `undefined` fields to be used as `CacheKey` values. Any fields containing `undefined` will be omitted during serialization.

## [6.0.0-beta.4] - 2020-08-25

### Changed

- Do not die when trying to cache errors with circular dependencies. However, we also no longer preserve custom properties on errors, because we don't know if they're safe to serialize.

## [6.0.0-beta.3] - 2020-08-25

### Added

- Added support for JSON objects as hash keys using `json-stable-stringify`.

## [6.0.0-beta.2] - 2020-08-24

### Changed

- Update to latest Redis client library.

## [6.0.0-beta.1] - 2020-08-23

### Changed

- Updated to TypeScript 4.
- Tests can now be run using `npm test`.

## [6.0.0-alpha.8] - 2020-08-19

### Added

- Full native TypeScript support.
- Linting using `eslint`.
- Code formatting using `prettier`.

### Changed

- The `cache-manager` library has been removed, and replaced with direct access to Redis and an LRU cache.
- Cache construction options have changed, if you choose to construct a cache manually (which is no longer encouraged).
- You should use `cache.wrap` and `cache.get` to acccess the cache.
- Locking has been totally rewritten.
- The test suite has been replaced with the excellent 4.x test suite, ported to `mocha`.

### Fixed

- Quite a few minor bugs have been fixed. Some new ones have probably been added.

### Removed

- A subtantial number of implementation details have been removed from the public API.


index.js
/**
 * Lock and cache using redis, memory, and more!
 * @example
 * let cache = lock_and_cache.LockAndCache()
 * let square = cache.wrap(async function foofunc (x) {return x*x})
 * const customCache = lock_and_cache.LockAndCache({opts})
 * let foo = customCache.get('foo', ()=>'bar')
 */

const { promisify } = require('util')
const crypto = require('crypto')
const redis = require('redis')
const cacheManager = require('cache-manager')
const redisStore = require('cache-manager-redis-store')
const ON_DEATH = require('death')

const inspect = require('util').inspect

function log (...message) {
  if (process.env.NODE_ENV === 'test' && !process.env.DEBUG) return
  message = [new Date(), ...message]
  console.log(...message.map((m) => {
    if (typeof m === 'string') return m.slice(0, 100)
    if (m instanceof Error) return m.stack
    if (m instanceof Date) return m.toLocaleString()
    return inspect(m, { colors: Boolean(process.stdout.isTTY) })
  }))
}

log.debug = function logDebug (...message) {
  if (process.env.NODE_ENV !== 'debug') return
  log(...message)
}

const LOCK_TIMEOUT = 60000
const LOCK_EXTEND_TIMEOUT = 5000

const ERROR_TTL = 1

const DEFAULT_REDIS_LOCK_OPTS = {
  retry_strategy (options) {
    return Math.min(options.attempt * 100, 3000)
  },
  url: (process.env.LOCK_URL || process.env.REDIS_URL || '//localhost:6379')
}

const DEFAULT_LOCK_OPTS = (
  process.env.NODE_ENV === 'test' ? {
    driftFactor: Number(process.env.LOCK_DRIFT_FACTOR) || null,
    retryCount: Number(process.env.LOCK_RETRY_COUNT) || 10,
    retryDelay: Number(process.env.LOCK_RETRY_DELAY) || 1000
  } : {
    driftFactor: Number(process.env.LOCK_DRIFT_FACTOR) || null,
    retryCount: Number(process.env.LOCK_RETRY_COUNT) || 36000,
    retryDelay: Number(process.env.LOCK_RETRY_DELAY) || 100
  }
)

const DEFAULT_MEM_CACHE_OPTS = {
  store: 'memory',
  max: 10,
  ttl: 30
}

const DEFAULT_REDIS_CACHE_OPTS = {
  store: redisStore,
  url: (process.env.CACHE_URL || process.env.REDIS_URL || '//localhost:6379'),
  ttl: 600
}

/**
 * runs cb() then calls obj.close()
 *
 * @example
 * closing(LockAndCache(), cache => console.log(cache.get('key', ()=>{return 'value'})))
 */
async function closing (obj, cb) {
  try {
    return await cb(obj)
  } finally {
    obj.close()
  }
}

/**
 * Create an array of mem/redis caches suitable for making a multi-tier cache;
 * passes opts directly to cache manager which passes directly to node redis
 * client.
 *
 * An attempt to provide useful defaults is made. These are provided by the
 * DEFAULT_MEM_CACHE_OPTS and DEFAULT_REDIS_CACHE_OPTS objects. User opts
 * override. Theoretically you could override the stores even.
 *
 * See also:
 *
 * - https://github.com/BryanDonovan/node-cache-manager
 * - https://github.com/dabroek/node-cache-manager-redis-store
 * - https://github.com/NodeRedis/node_redis
 *
 * @example
 * tieredCache()
 * @example
 * tieredCache({max: 20, ttl: 60}, {ttl: 1200})
 */
function tieredCache (memOpts, redisOpts) {
  return [
    cacheManager.caching(Object.assign({}, DEFAULT_MEM_CACHE_OPTS, memOpts)), cacheManager.caching(Object.assign({}, DEFAULT_REDIS_CACHE_OPTS, redisOpts)) ]
}

class KeyNotFoundError extends Error {}

class AsyncRedis {
  constructor (client) {
    this.get = promisify(client.get).bind(client)
    this.set = promisify(client.set).bind(client)
    this.del = promisify(client.del).bind(client)
    this.eval = promisify(client.eval).bind(client)
    this.quit = promisify(client.quit).bind(client)
    this.flushall = promisify(client.flushall).bind(client)
  }
}

class LockError extends Error {}
class LockExtendError extends LockError {}
class LockUnlockError extends LockError {}

class Lock {
  constructor (asyncRedisClient, key, secret, timeout) {
    this._client = asyncRedisClient
    this.key = key
    this.secret = secret
    this.timeout = timeout
    this.done = true
    this.extendErr = null
    this.extendTimeoutHandle = null
    this.lastExtension = null
    this.extension = Promise.resolve()
    this.locked = true
    this.extendMissedDeadlines = 0
  }

  extendForever () {
    if (this.done && !this.extendErr) {
      this.done = false
      this.lastExtension = new Date()
      this._extendForeverLoop()
    } else {
      throw this.extendErr
    }
  }

  async _extendForeverLoop () {
    try {
      if (!this.done) {
        this.extension = this.extend()
        await this.extension
        const timestamp = new Date()
        const latency = timestamp - this.lastExtension
        this.lastExtension = timestamp
        if (latency > LOCK_EXTEND_TIMEOUT * 1.1) {
          ++this.extendMissedDeadlines
          log(`warning: missed lock extension deadline by ${latency - LOCK_EXTEND_TIMEOUT}ms; ${this.extendMissedDeadlines} total misses`)
        }
        // scheduler overhead fudge factor
        const SCHED_FUDGE = 100
        const ERR_FACTOR = Math.max(0, latency - LOCK_EXTEND_TIMEOUT)
        if (!this.done) {
          const timeout = Math.max(0, LOCK_EXTEND_TIMEOUT - SCHED_FUDGE - ERR_FACTOR)
          this.extendTimeoutHandle = setTimeout(this._extendForeverLoop.bind(this), timeout)
        }
      }
    } catch (err) {
      this.done = true
      this.extendErr = err
    }
  }

  async _stopExtendLoop () {
    this.done = true
    clearTimeout(this.extendTimeoutHandle)
    await this.extension
  }

  async unlock () {
    if (!this.locked) {
      throw new LockUnlockError('not locked')
    }
    log.debug(`unlock ${this.key}, ${this.secret}`)
    this.locked = false
    await this._stopExtendLoop()
    const r = await this._client.eval(
      `if redis.call("get", KEYS[1]) == ARGV[1] then
        return redis.call("del", KEYS[1])
      else
        if redis.call("get", KEYS[1]) == false then
          return -1
        else
          return -2
        end
      end`,
      1, this.key, this.secret
    )
    if (r === 0) {
      throw new LockUnlockError('del failed; this should NEVER happen')
    }
    if (r === -1) {
      throw new LockUnlockError('lock expired')
    }
    if (r === -2) {
      throw new LockUnlockError('lock expired and was taken')
    }
  }

  async extend () {
    if (!this.locked) {
      throw new LockExtendError('not locked')
    }
    log.debug(`extend ${this.key}, ${this.secret}`)
    const r = await this._client.eval(
      `if redis.call("get", KEYS[1]) == ARGV[1] then
        return redis.call("pexpire", KEYS[1], ARGV[2])
      else
        if redis.call("get", KEYS[1]) == false then
          return -1
        else
          return -2
        end
      end`,
      1, this.key, this.secret, this.timeout
    )
    if (r === 0) {
      throw new LockExtendError('pexpire failed; this should NEVER happen')
    }
    if (r === -1) {
      throw new LockExtendError('lock expired')
    }
    if (r === -2) {
      throw new LockExtendError('lock expired and was taken')
    }
  }
}

class FailedToLock extends Error {}

class LockManager {
  constructor (asyncRedisClient) {
    this._client = asyncRedisClient
  }

  async lockRetryExtending (key) {
    const lock = await this.lockRetry(key)
    lock.extendForever()
    return lock
  }

  async lockRetry (key) {
    while (true) {
      try {
        return await this.lock(key)
      } catch (err) {
        if (err instanceof FailedToLock) {
          await new Promise(resolve => setTimeout(resolve, LOCK_EXTEND_TIMEOUT))
        } else {
          throw err
        }
      }
    }
  }

  async lock (key) {
    const secret = crypto.randomBytes(64).toString('base64')
    const _key = 'lock:' + key
    log.debug('try lock', _key, secret)
    if (!await this._client.set(_key, secret, 'nx', 'px', LOCK_TIMEOUT)) {
      throw new FailedToLock('failed to aquire lock; already locked')
    }
    log.debug('locked', _key, secret)
    return new Lock(this._client, _key, secret, LOCK_TIMEOUT)
  }
}

/**
 * LockAndCache
 *
 * By default, LockAndCache caches values; that is to say it deep-clones objects
 * prior to cache insertion and subsequent to cache retrieval. This prevents
 * bugs due to side-effects of mutation of cached data.
 *
 * This behavior can be changed by setting `byReference:true` in the
 * constructor. Only memory stores correctly support byReference because other
 * stores must necessarily serialize objects. The performance gain will depend
 * on the size of the objects and will usually be small, but if the data is
 * immutable then there is no reason to copy it. There are other reasons why you
 * might want to cache the original objects, e.g. if they are difficult or
 * impossible to serialize, e.g. class instances.
 *
 * @example
 * const cache = LockAndCache()
 * try {
 *   const val = cache.get('key', async function () { return 'value' })
 * } finally {
 *   cache.close()
 * }
 *
 *
 */
class LockAndCache {
  /**
   * LockAndCache constructor
   * @param {Array} [caches=tieredCache()]                                      Array of caches to be passed to cacheManager.multiCaching
   * @param {Array}  [lockClients=[redis.createClient(DEFAULT_REDIS_LOCK_OPTS)]] Array of redis clients for redlock
   * @param {Object} [lockOpts=DEFAULT_LOCK_OPTS]                             Options to pass to redlock constructor
   * @param {Boolean} [byReference=false]                                                        } = {}] Cache by reference instead of value
   */
  constructor ({
    caches = tieredCache(),
    lockClient = redis.createClient(DEFAULT_REDIS_LOCK_OPTS),
    lockOpts = DEFAULT_LOCK_OPTS,
    byReference = false
  } = {}) {
    this._byReference = byReference
    this._lockClient = lockClient
    this._asyncLockClient = new AsyncRedis(lockClient)
    this._lockManager = new LockManager(this._asyncLockClient)
    this._cacheClients = (
      caches.map(cache => cache.store.getClient && cache.store.getClient())
        .filter(cache => !!cache)
    )
    this._cache = cacheManager.multiCaching(caches)
    this.OFF_DEATH = ON_DEATH(() => this.close())
  }

  _cacheGetTransform (value) {
    if (value === 'undefined') return undefined
    if (this._byReference) {
      return value
    }
    return JSON.parse(value)
  }

  _stringifyKey (key) {
    return typeof key === 'string' ? key : JSON.stringify(key)
  }

  async _cacheGet (key) {
    key = this._stringifyKey(key)
    let value = await this._cache.get(key)
    if (value === null || typeof value === 'undefined') {
      throw new KeyNotFoundError(key)
    }
    value = this._cacheGetTransform(value)
    // log.debug('got', JSON.stringify(value).slice(100), 'for', key)
    return value
  }

  _cacheSetTransform (value) {
    if (typeof value === 'undefined') return 'undefined'
    if (this._byReference) {
      return value
    }
    return JSON.stringify(value)
  }

  async _cacheSet (key, value, ttl) {
    key = this._stringifyKey(key)
    let v = await this._cache.set(key, this._cacheSetTransform(value), {
      ttl
    })
    // log.debug('set', value.toString().slice(100), 'for', key)
    return v
  }

  close () {
    log.debug('closing connections', 1 + this._cacheClients.length)
    this.OFF_DEATH();
    [this._lockClient, ...this._cacheClients].forEach(c => c.quit())
  }

  del (...opts) {
    return this._cache.del(...opts)
  }

  _errorFactory (err) {
    let serializedError
    if (err instanceof Error) {
      serializedError = {}
      if (err.name) serializedError.name = err.name
      if (err.message) serializedError.message = err.message
      if (err.stack) serializedError.stack = err.stack
      Object.assign(serializedError, err) // retain custom properties
    } else {
      serializedError = String(err)
    }
    return {
      __lock_and_cache_error__: true,
      err: serializedError
    }
  }

  // this is called "get" to match up with standard cache library semantics
  // but don't forget it also locks
  async get (key, ttl, work) {
    let value
    key = this._stringifyKey(key)

    log.debug('get', key)

    if (typeof work === 'undefined') {
      work = ttl
      ttl = undefined
    }

    try {
      return await this._cacheGet(key)
    } catch (err) {
      if (!(err instanceof KeyNotFoundError)) throw err
    }

    const lock = await this._lockManager.lockRetryExtending(key)

    try {
      try {
        return await this._cacheGet(key)
      } catch (err) {
        if (!(err instanceof KeyNotFoundError)) throw err
      }
      try {
        log.debug('calling work to compute value')
        value = await work()
        return value
      } catch (err) {
        ttl = ERROR_TTL
        value = this._errorFactory(err)
        if (typeof value === 'object' && value !== null && value.__lock_and_cache_error__) {
          if (typeof value.err === 'object') {
            const err = new Error()
            Object.assign(err, value.err)
            throw err
          } else {
            throw new Error(value.err)
          }
        }
      } finally {
        await lock.extend()
        await this._cacheSet(key, value, ttl)
      }
    } catch (err) {
      if (lock.extendErr) {
        log(lock.extendErr)
      }
      throw err
    } finally {
      await lock.unlock()
    }
    if (lock.extendErr) {
      throw lock.extendErr
    }
  }

  /**
   * wraps a function in a locking cache
   * @param  {string|number|function} name if 3 params, ttl if 2 params, work if 1 param
   * @param  {number|function} ttl  if 3 params, work if 2 params
   * @param  {function} work if 3 params
   * @return {function}      wrapped function
   */
  wrap (...opts) {
    let name, ttl, work
    if (opts.length === 3) {
      [name, ttl, work] = opts
    } else if (opts.length === 2) {
      [ttl, work] = opts
    } else if (opts.length === 1) {
      [work] = opts
    } else {
      throw new TypeError('wrap requires 1, 2, or 3 arguments')
    }

    if (typeof name !== 'string') {
      name = work.displayName || work.name
    }

    log.debug('wrap', name)

    if (!name) {
      throw new TypeError('lockAndCache.wrap(work) requires named function')
    }

    if (typeof name !== 'string') throw new TypeError('name must be a string')
    if (typeof work !== 'function') throw new TypeError('work must be a function')

    const wrappedFn = async function (...args) {
      log.debug('call wrapped', name, ...args)
      const key = [name].concat(args)
      return this.get(key, ttl, async function doWork () {
        return work(...args)
      })
    }.bind(this)
    wrappedFn.displayName = name

    return wrappedFn
  }
}

module.exports.LockAndCache = LockAndCache
module.exports.tieredCache = tieredCache
module.exports.closing = closing
module.exports.DEFAULT_MEM_CACHE_OPTS = DEFAULT_MEM_CACHE_OPTS
module.exports.DEFAULT_REDIS_CACHE_OPTS = DEFAULT_REDIS_CACHE_OPTS
module.exports.DEFAULT_REDIS_LOCK_OPTS = DEFAULT_REDIS_LOCK_OPTS
module.exports.KeyNotFoundError = KeyNotFoundError
module.exports.LOCK_EXTEND_TIMEOUT = LOCK_EXTEND_TIMEOUT
module.exports.Lock = Lock
module.exports.AsyncRedis = AsyncRedis
